proposed/2021/BulkRestoreCoordination.md

```diff
@@ -0,0 +1,170 @@
+# Bulk restore coordination
+
+- [Nikolche Kolev](https://github.com/nkolev92)
+- Start Date 2021-05-03
+- GitHub Issue [10678](https://github.com/NuGet/Home/issues/10678)
+- GitHub PR [4058](https://github.com/NuGet/NuGet.Client/pull/4058)
+
+## Summary
+
+<!-- One-paragraph description of the proposal. -->
+When a solution with SDK based projects is loaded, NuGet uses a sliding window to coordinate the number of restores being run to avoid extra restore/design time build loops.
+When a solution is loaded, any future changes will trigger a restore eagerly. Operations such as branch switch are one example.
+This proposal tackles the addition of a new API to coordinate the restore during the bulk changes operations.
+
+## Motivation
+
+<!-- Why are we doing this? What pain points does this solve? What is the expected outcome? -->
+
+There are a few scenarios where a bulk project edit may happen that affects NuGet restore.
+
+- Branch switch
+- Directory.Build.props or other shared file edits
+- Find and replace in csproj
+
+These operations are currently not very efficient and may run extra restores and design time builds. In some cases, re-opening a solution may be more efficient than a partial branch switch.
+
+## Explanation
+
+### Functional explanation
+
+<!-- Explain the proposal as if it were already implemented and you're teaching it to another person. -->
+<!-- Introduce new concepts, functional designs with real life examples, and low-fidelity mockups or  pseudocode to show how this proposal would look. -->
+
+There are no functional changes to any experiences.
+The user actions are all the same, the changes have a performance focus.
+
+### Technical explanation
+
+<!-- Explain the proposal in sufficient detail with implementation details, interaction models, and clarification of corner cases. -->
+#### Target scenario
+
+Ex:
+Given a solution with 3 projects.
+Proj A -> Proj B -> Proj C.
+
+Say a Directory.Build.Props is edited that affect *all* projects.
+Say the nominations come in the following order:
+
+Proj B
+Proj C
+Proj A
+
+- When the nomination for B comes in, NuGet will trigger a restore that will affect projects B & A, given that A depends on B.

```

Comment:
Last question. Here nomination means something changed so CPS is telling VS this project is ready for restore. Is it right?

Reply:
Yes, mostly.

I am going to nitpick your statement an attempt to clearly communicate the separation. 

> CPS is telling VS this project is ready for restore

Firstly:

> CPS is telling VS

CPS is really the project-system. In theory it should be CPS that's calling NuGet, but it's actually managed languages project system right now.

> this project is ready for restore

There's a model project system that's generated from design time builds. When a design time build is run, the data is collected and compared against the last data that project-system sent to NuGet. 
If it is different, a nomination call happens. 

Each nomination call tells NuGet that a restore *may* be needed. It's NuGet that decides whether a restore runs and if it does whether any changes occur. 

Hope that clarifies it.